src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,378 @@
+    #         self.consume_stream_datum(*stream_datum_docs)
+    #         import h5py
+    #
+    #         array = ...
+    #         return array
+
+
+class TIFFStreamHandler(StreamHandlerBase):
+    mimetype = "multipart/related;type=image/tiff"
+
+    def __init__(self, data_uri, data_shape, data_type, **kwargs):
+        self.data_uri = data_uri
+        self.assets = []
+        self.data_shape = data_shape
+        self.data_type = np.dtype(data_type)
+        self.adapter_parameters = {"data_uris": []}
+
+    @property
+    def shape(self):
+        """Shape including the 0-th dimension -- number of rows"""
+        # TODO Add reshape, skips.
+        return [self._num_rows, *self.data_shape]
+
+    def consume_stream_datum(self, doc):
+        indx = int(doc["uid"].split("/")[1])
+        new_data_uri = self.data_uri.strip("/") + "/" + f"{indx:05d}.tif"
+        new_asset = Asset(
+            data_uri=new_data_uri,
+            is_directory=False,
+            parameter="data_uris",
+            num=len(self.assets) + 1,
+        )
+        self.assets.append(new_asset)
+        self.adapter_parameters["data_uris"].append(new_data_uri)
+        self._num_rows += doc["indices"]["stop"] - doc["indices"]["start"]  # Number of rows added by StreamDatum
+
+    def get_data_source(self):
+        return DataSource(
+            mimetype=self.mimetype,
+            assets=self.assets,
+            structure_family=StructureFamily.array,
+            structure=ArrayStructure(
+                data_type=BuiltinDtype.from_numpy_dtype(self.data_type),
+                shape=self.shape,
+                chunks=[[1] * self.shape[0]] + [[d] for d in self.data_shape],
+            ),
+            parameters=self.adapter_parameters,
+            management=Management.external,
+        )
+
+
+STREAM_HANDLER_REGISTRY = {
+    "application/x-hdf5": HDF5StreamHandler,
+    "multipart/related;type=image/tiff": TIFFStreamHandler,
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self.root_node = None
+        self._desc_nodes = {}  # references to descriptor containers by their uid's
+        self._sr_nodes = {}
+        self._sr_cache = {}
+        self._handlers = {}
+
```

Comment:
These would benefit from type hints, as would the change generally: is _desc_nodes presumably a dict[Descriptor, list[Event]]?

Reply:
Fixed this and added more type hints throughout.